internal/common/customplanmodifier/plan_modify_differ.go

```diff
@@ -0,0 +1,303 @@
+		}
+	}
+	return unknowns
+}
+
+func (d *PlanModifyDiffer) UseStateForUnknown(ctx context.Context, diags *diag.Diagnostics, keepUnknown []string, prefix path.Path) {
+	// The diff is sorted by the path length, for example read_only_spec is processed before read_only_spec.disk_size_gb
+	schema := d.schema
+	for _, diff := range d.statePlanDiff {
+		stateValue, tpfPath := conversion.AttributePathValue(ctx, diags, diff.Path, d.req.State, schema)
+		if !conversion.HasPrefix(tpfPath, prefix) || stateValue == nil || conversion.IsIndexValue(tpfPath) {
+			continue
+		}
+		if d.ParentRemoved(tpfPath) {
+			continue
+		}
+		planValue, _ := conversion.AttributePathValue(ctx, diags, diff.Path, d.req.Plan, schema)
+		if planValue == nil || !planValue.IsUnknown() {
+			continue
+		}
+		if keepUnknownCall(diff.Path, keepUnknown) {
+			tflog.Info(ctx, fmt.Sprintf("Keeping unknown value in plan @ %s", tpfPath.String()))
+			unknownValue := conversion.AsUnknownValue(ctx, stateValue)
+			UpdatePlanValue(ctx, diags, d, tpfPath, unknownValue)
+		} else {
+			tflog.Info(ctx, fmt.Sprintf("Replacing unknown value in plan @ %s", tpfPath.String()))
+			UpdatePlanValue(ctx, diags, d, tpfPath, stateValue)
+			d.ensureKeepUnknownRespected(ctx, diags, tpfPath, stateValue, keepUnknown)
+		}
+	}
+}
+
+func (d *PlanModifyDiffer) ensureKeepUnknownRespected(ctx context.Context, diags *diag.Diagnostics, tpfPath path.Path, value attr.Value, keepUnknown []string) {
+	valueObject, ok := value.(types.Object)
+	if value.IsNull() || value.IsUnknown() || !ok {
+		return
+	}
+	for key, childValue := range valueObject.Attributes() {
+		if slices.Contains(keepUnknown, key) && !childValue.IsUnknown() {
+			childPath := tpfPath.AtName(key)
+			tflog.Info(ctx, fmt.Sprintf("Keeping unknown value in plan @ %s", childPath.String()))
+			unknownValue := conversion.AsUnknownValue(ctx, childValue)
+			UpdatePlanValue(ctx, diags, d, childPath, unknownValue)
+		}
+	}
+}
+
+func ReadConfigStructValue[T any](ctx context.Context, d *PlanModifyDiffer, p path.Path) *T {
+	return readSrcStructValue[T](ctx, d.req.Config, p)
+}
+
+func ReadPlanStructValue[T any](ctx context.Context, d *PlanModifyDiffer, p path.Path) *T {
+	return readSrcStructValue[T](ctx, d.req.Plan, p)
+}
+
+func ReadStateStructValue[T any](ctx context.Context, d *PlanModifyDiffer, p path.Path) *T {
+	return readSrcStructValue[T](ctx, d.req.State, p)
+}
+
+func readSrcStructValue[T any](ctx context.Context, src conversion.TPFSrc, p path.Path) *T {
+	var obj types.Object
+	if localDiags := src.GetAttribute(ctx, p, &obj); localDiags.HasError() {
+		return nil
+	}
+	if obj.IsNull() || obj.IsUnknown() {
+		return nil
+	}
+	return conversion.TFModelObject[T](ctx, obj)
+}
+
+func UpdatePlanValue[T attr.Value](ctx context.Context, diags *diag.Diagnostics, d *PlanModifyDiffer, p path.Path, value T) {
+	if localDiags := d.resp.Plan.SetAttribute(ctx, p, value); localDiags.HasError() {
+		diags.Append(localDiags...)
+	}
+}
```

Comment:
```suggestion
func UpdatePlanValue(ctx context.Context, diags *diag.Diagnostics, d *PlanModifyDiffer, p path.Path, value attr.Value) {
	diags.Append(d.resp.Plan.SetAttribute(ctx, p, value)...)
		diags.Append(localDiags...)
	}
}
```

no need for generics as it's an interface, copy also warnings if any

Reply:
Thanks changed in 77eb77b81d993280f144c9c4c59b936b944c2c3c